cpp/include/raft/detail/serialization.hpp

```diff
@@ -0,0 +1,205 @@
+  {
+    if (out) { *reinterpret_cast<T*>(out) = obj; }
+    return sizeof(S);
+  }
+
+  // SFINAE-style failure
+  template <typename S = T, typename... ContextArgs>
+  static auto to_bytes(uint8_t* out, const S& obj, ContextArgs&&... args)
+    -> std::enable_if_t<!std::is_arithmetic_v<S>, size_t>
+  {
+    static_assert(!std::is_same_v<T, S>, "Serialization is not implemented for this type.");
+    return 0;
+  }
+
+  // Default implementation for all arithmetic types: just read the value by the pointer.
+  template <typename S = T>
+  static auto from_bytes(S* p, const uint8_t* in)
+    -> std::enable_if_t<std::is_arithmetic_v<S>, size_t>
+  {
+    *p = *reinterpret_cast<const S*>(in);
+    return sizeof(S);
+  }
+
+  // SFINAE-style failure
+  template <typename S = T, typename... ContextArgs>
+  static auto from_bytes(S* p, const uint8_t* in, ContextArgs&&... args)
+    -> std::enable_if_t<!std::is_arithmetic_v<S>, size_t>
+  {
+    static_assert(!std::is_same_v<T, S>, "Deserialization is not implemented for this type.");
+    return 0;
+  }
+};
+
+template <typename IndexType, size_t... ExtentsPack>
+struct serial<extents<IndexType, ExtentsPack...>> {
+  using obj_t = extents<IndexType, ExtentsPack...>;
+
+  static auto to_bytes(uint8_t* out, const obj_t& obj) -> size_t
+  {
+    if (out) { *reinterpret_cast<obj_t*>(out) = obj; }
+    return sizeof(obj_t);
+  }
+
+  static auto from_bytes(obj_t* p, const uint8_t* in) -> size_t
+  {
+    new (p) obj_t{*reinterpret_cast<const obj_t*>(in)};
+    return sizeof(obj_t);
+  }
+};
+
+template <typename ElementType, typename Extents, typename LayoutPolicy>
+struct serial<mdarray<ElementType,
+                      Extents,
+                      LayoutPolicy,
+                      detail::device_accessor<detail::device_uvector_policy<ElementType>>>> {
+  using obj_t = mdarray<ElementType,
+                        Extents,
+                        LayoutPolicy,
+                        detail::device_accessor<detail::device_uvector_policy<ElementType>>>;
+
+  static auto to_bytes(uint8_t* out, const obj_t& obj, const handle_t& handle) -> size_t
+  {
+    auto extents_size = call_serialize<Extents>(out, obj.extents());
+    auto total_size   = obj.size() * sizeof(ElementType) + extents_size;
+    if (out) {
+      out += extents_size;
+      raft::copy(
+        reinterpret_cast<ElementType*>(out), obj.data_handle(), obj.size(), handle.get_stream());
+    }
+    return total_size;
+  }
+
+  static auto from_bytes(obj_t* p,
+                         const uint8_t* in,
+                         const handle_t& handle,
+                         rmm::mr::device_memory_resource* mr = nullptr) -> size_t
+  {
+    Extents exts;
+    auto extents_size = call_deserialize<Extents>(&exts, in);
+    in += extents_size;
+    typename obj_t::mapping_type layout{exts};
+    typename obj_t::container_policy_type policy{handle.get_stream(), mr};
+    new (p) obj_t{layout, policy};
+    raft::copy(
```

Comment:
[INCORRECT]

This is only correct if the layout is exhaustive (contiguous in memory, not strided or otherwise skipping parts of memory).  Please see above note.